server/src/models/mod.rs

```diff
@@ -12,134 +12,263 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+//! Implementation of the model types used for compressing time series segments
+//! as models and functions for efficiently computing aggregates from models of
+//! each type. The module itself contains functionality used by multiple of the
+//! model types.
+
 mod gorilla;
 mod pmcmean;
 mod swing;
 
-use datafusion::arrow::array::{
-    Float32Builder, Int32Array, Int32Builder, Int64Array, TimestampMillisecondBuilder,
+use std::cmp::{Ordering, PartialOrd};
+
+use datafusion::arrow::array::{Int32Array, Int64Array};
+
+use crate::errors::MiniModelarDBError;
+use crate::types::{
+    TimeSeriesId, TimeSeriesIdArray, TimeSeriesIdBuilder, Timestamp, TimestampBuilder, Value,
+    ValueBuilder,
 };
 
-// TODO: can mtid be converted to a model type enum without adding overhead?
+/// Unique ids for each model type. Constant values are used instead of an enum
+/// so the stored model type ids can be used in match expressions without being
+/// converted to an enum first.
+const PMC_MEAN_ID: u8 = 2;
+const SWING_ID: u8 = 3;
+const GORILLA_ID: u8 = 4;
+
+/// General error bound that is guaranteed to not be negative, infinite, or NAN.
+/// For `PMCMean` and `Swing` the error bound is interpreted as a relative per
+/// value error bound in percentage, while `Gorilla` uses lossless compression.
+struct ErrorBound(f32);
+
+impl ErrorBound {
+    /// Return `Self` if `error_bound` is a positive finite value, otherwise
+    /// `CompressionError`.
+    fn try_new(error_bound: f32) -> Result<Self, MiniModelarDBError> {
+        if error_bound < 0.0 || error_bound.is_infinite() || error_bound.is_nan() {
+            Err(MiniModelarDBError::CompressionError(
+                "Error bound cannot be negative, infinite, or NAN".to_owned(),
```

Comment:
[`to_owned()`](https://doc.rust-lang.org/std/borrow/trait.ToOwned.html) is [generally recommended](https://users.rust-lang.org/t/to-string-vs-to-owned-for-string-literals/1441) because it can be optimized to a higher degree than [`to_string()`](https://doc.rust-lang.org/std/string/trait.ToString.html). So while both methods works since [`str`](https://doc.rust-lang.org/std/primitive.str.html) implements both traits and I think `to_string()` is more intuitive, I suggest we use `to_owned()` and have removed a leftover `to_string()` from the models module.

Reply:
Sounds like a good idea, I will change the storage engine to do this as well.